### ISSUE TITLE
upgrading v0.1 to v0.2, reading "old" MPS/ITensor format

### ISSUE THREAD

**Issue by jtschneider** [ITensor/ITensors.jl#714]
Hello guys!

I have the following issue: I would like to convert an old MPS / ITensor saved on my disk (inside a HDF5 file) to the new type. Is there a way around downgrading to v0.1 when reading the old HDF5 files?
Could we build a function converting the old ITensor type to the new one?
Just wondering how I can save my computations while still having the bug fixes and features of v0.2 :thinking:

**Comment by mtfishman** [comment]
So the issue is that you had saved an MPS to disk using HDF5 using ITensors v0.1, but now you are having trouble reading that same file when using ITensors v0.2? Any chance you could provide a minimal example of the issue?

**Comment by mtfishman** [comment]
All of the data within the HDF5 file should be the same in ITensors v0.1 and v0.2 so in principle I don't think there should be an issue getting this to work, so in practice I am curious what issue you see.

**Comment by jtschneider** [comment]
> So the issue is that you had saved an MPS to disk using HDF5 using ITensors v0.1, but now you are having trouble reading that same file when using ITensors v0.2? Any chance you could provide a minimal example of the issue?

Exactly! I have an MPS computed with ITensors v0.1 saved to disk but when I try to open it with ITensors v0.2 it does not work.

With regards of a simple example, let me try with this little construction; we need two test paths, where we initialise a project with two different versions of ITensors.jl:

```zsh
cd ~/Documents/Code/
mkdir testenv1
mkdir testenv2
```

So I change dir to the respective folder, start julia, and activate the local project there (either by `julia --project` or by `julia` and then `] activate .` ) 

```julia
julia> ]

(@v1.6) pkg> activate .
  Activating new environment at `~/Documents/Code/testenv1/Project.toml`

(testenv1) pkg> status
      Status `~/Documents/Code/testenv1/Project.toml` (empty project)

(testenv1) pkg> add ITensors@v0.1.41
    Updating registry at `~/.julia/registries/General`
   Resolving package versions...
   Installed PackageCompiler ─ v1.3.0
   Installed ITensors ──────── v0.1.41
   Installed NDTensors ─────── v0.1.28
    Updating `~/Documents/Code/testenv1/Project.toml`
  [9136182c] + ITensors v0.1.41
    Updating `~/Documents/Code/testenv1/Manifest.toml`
  [... download all dependencies with correct version...]
Precompiling project...
  3 dependencies successfully precompiled in 29 seconds (22 already precompiled)

julia> using ITensors, ITensors.HDF5, Random

julia> Random.seed!(42)
MersenneTwister(42)

julia> N = 12
12

julia> sites = siteinds("S=1",N)
12-element Vector{Index{Int64}}:
 (dim=3|id=462|"S=1,Site,n=1")
 (dim=3|id=359|"S=1,Site,n=2")
 (dim=3|id=324|"S=1,Site,n=3")
 (dim=3|id=731|"S=1,Site,n=4")
 (dim=3|id=216|"S=1,Site,n=5")
 (dim=3|id=243|"S=1,Site,n=6")
 (dim=3|id=271|"S=1,Site,n=7")
 (dim=3|id=119|"S=1,Site,n=8")
 (dim=3|id=94|"S=1,Site,n=9")
 (dim=3|id=843|"S=1,Site,n=10")
 (dim=3|id=946|"S=1,Site,n=11")
 (dim=3|id=331|"S=1,Site,n=12")

julia> # Create an initial random matrix product state
         psi0 = randomMPS(sites,10)
MPS
[1] IndexSet{2} (dim=3|id=462|"S=1,Site,n=1") (dim=10|id=368|"Link,l=1")
[2] IndexSet{3} (dim=10|id=368|"Link,l=1") (dim=3|id=359|"S=1,Site,n=2") (dim=10|id=65|"Link,l=2")
[3] IndexSet{3} (dim=10|id=65|"Link,l=2") (dim=3|id=324|"S=1,Site,n=3") (dim=10|id=387|"Link,l=3")
[4] IndexSet{3} (dim=10|id=387|"Link,l=3") (dim=3|id=731|"S=1,Site,n=4") (dim=10|id=307|"Link,l=4")
[5] IndexSet{3} (dim=10|id=307|"Link,l=4") (dim=3|id=216|"S=1,Site,n=5") (dim=10|id=962|"Link,l=5")
[6] IndexSet{3} (dim=10|id=962|"Link,l=5") (dim=3|id=243|"S=1,Site,n=6") (dim=10|id=313|"Link,l=6")
[7] IndexSet{3} (dim=10|id=313|"Link,l=6") (dim=3|id=271|"S=1,Site,n=7") (dim=10|id=194|"Link,l=7")
[8] IndexSet{3} (dim=10|id=194|"Link,l=7") (dim=3|id=119|"S=1,Site,n=8") (dim=10|id=176|"Link,l=8")
[9] IndexSet{3} (dim=10|id=176|"Link,l=8") (dim=3|id=94|"S=1,Site,n=9") (dim=10|id=277|"Link,l=9")
[10] IndexSet{3} (dim=10|id=277|"Link,l=9") (dim=3|id=843|"S=1,Site,n=10") (dim=9|id=582|"Link,l=10")
[11] IndexSet{3} (dim=9|id=582|"Link,l=10") (dim=3|id=946|"S=1,Site,n=11") (dim=3|id=835|"Link,l=11")
[12] IndexSet{2} (dim=3|id=835|"Link,l=11") (dim=3|id=331|"S=1,Site,n=12")

julia> h5open("testfilev0.1.h5", "w") do file 
           write(file, "psi0v0.1", psi0)
       end
```

Now, I am using the newer version of ITensors;

```julia
julia> cd("~/Documents/Code/testenv2")

julia> ]

(@v1.6) pkg> activate .
  Activating new environment at `~/Documents/Code/testenv2/Project.toml`

(testenv2) pkg> add ITensors
    Updating registry at `~/.julia/registries/General`
   Resolving package versions...
    Updating `~/Documents/Code/testenv2/Project.toml`
  [9136182c] + ITensors v0.2.4
    Updating `~/Documents/Code/testenv2/Manifest.toml`

(testenv2) pkg> status
      Status `~/Documents/Code/testenv2/Project.toml`
  [9136182c] ITensors v0.2.4

julia> using ITensors, ITensors.HDF5, Random
 
julia> N = 12
12

julia> file01 = h5open("../testenv1/testfilev0.1.h5", "r");


julia> psi0_01 = read(file01,"psi0v0.1", MPS)
HDF5-DIAG: Error detected in HDF5 (1.12.0) thread 0:
  #000: H5G.c line 528 in H5Gopen2(): unable to open group
    major: Symbol table
    minor: Can't open object
  #001: H5VLcallback.c line 4186 in H5VL_group_open(): group open failed
    major: Virtual Object Layer
    minor: Can't open object
  #002: H5VLcallback.c line 4153 in H5VL__group_open(): group open failed
    major: Virtual Object Layer
    minor: Can't open object
  #003: H5VLnative_group.c line 125 in H5VL__native_group_open(): unable to open group
    major: Symbol table
    minor: Can't open object
  #004: H5Gint.c line 282 in H5G__open_name(): group not found
    major: Symbol table
    minor: Object not found
  #005: H5Gloc.c line 462 in H5G_loc_find(): can't find object
    major: Symbol table
    minor: Object not found
  #006: H5Gtraverse.c line 855 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #007: H5Gtraverse.c line 630 in H5G__traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #008: H5Gloc.c line 418 in H5G__loc_find_cb(): object 'storage' doesn't exist
    major: Symbol table
    minor: Object not found
ERROR: Error opening group /psi0v0.1/MPS[1]/storage
Stacktrace:
 [1] error(::String, ::String, ::String, ::String)
   @ Base ./error.jl:42
 [2] h5g_open
   @ ~/.julia/packages/HDF5/0iEnL/src/api.jl:702 [inlined]
 [3] open_group(parent::HDF5.Group, name::String, apl::HDF5.Properties) (repeats 2 times)
   @ HDF5 ~/.julia/packages/HDF5/0iEnL/src/HDF5.jl:659
 [4] read(parent::HDF5.Group, name::String, #unused#::Type{ITensor})
   @ ITensors ~/.julia/packages/ITensors/MATKY/src/itensor.jl:2568
 [5] #840
   @ ./none:0 [inlined]
 [6] iterate
   @ ./generator.jl:47 [inlined]
 [7] collect(itr::Base.Generator{UnitRange{Int64}, ITensors.var"#840#841"{HDF5.Group}})
   @ Base ./array.jl:678
 [8] read(parent::HDF5.File, name::String, #unused#::Type{MPS})
   @ ITensors ~/.julia/packages/ITensors/MATKY/src/mps/mps.jl:736
 [9] top-level scope
   @ REPL[10]:1
```

It does however work to read the MPS as a dictionary:
```julia
julia> read(file01,"psi0v0.1")
Dict{String, Any} with 15 entries:
  "MPS[9]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[10]" => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[1]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>2, "index_1"=>Dict{String, Any}(…
  "MPS[3]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[4]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[11]" => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[6]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "length"  => 12
  "MPS[12]" => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>2, "index_1"=>Dict{String, Any}(…
  "MPS[8]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "llim"    => 0
  "MPS[2]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[7]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[5]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "rlim"    => 2
```

**Comment by jtschneider** [comment]
In fact, reading the HDF5 file as above gives a hint where to look for the incompatibility:

Given above execution history in `testenv2`, we can test the following:

```julia
julia> N = 12
12

julia> sites = siteinds("S=1",N)
12-element Vector{Index{Int64}}:
 (dim=3|id=124|"S=1,Site,n=1")
 (dim=3|id=612|"S=1,Site,n=2")
 (dim=3|id=294|"S=1,Site,n=3")
 (dim=3|id=683|"S=1,Site,n=4")
 (dim=3|id=616|"S=1,Site,n=5")
 (dim=3|id=427|"S=1,Site,n=6")
 (dim=3|id=909|"S=1,Site,n=7")
 (dim=3|id=334|"S=1,Site,n=8")
 (dim=3|id=947|"S=1,Site,n=9")
 (dim=3|id=611|"S=1,Site,n=10")
 (dim=3|id=564|"S=1,Site,n=11")
 (dim=3|id=700|"S=1,Site,n=12")

julia> psi0 = randomMPS(sites,10)
MPS
[1] ((dim=3|id=124|"S=1,Site,n=1"), (dim=10|id=567|"Link,l=1"))
[2] ((dim=10|id=567|"Link,l=1"), (dim=3|id=612|"S=1,Site,n=2"), (dim=10|id=46|"Link,l=2"))
[3] ((dim=10|id=46|"Link,l=2"), (dim=3|id=294|"S=1,Site,n=3"), (dim=10|id=147|"Link,l=3"))
[4] ((dim=10|id=147|"Link,l=3"), (dim=3|id=683|"S=1,Site,n=4"), (dim=10|id=129|"Link,l=4"))
[5] ((dim=10|id=129|"Link,l=4"), (dim=3|id=616|"S=1,Site,n=5"), (dim=10|id=541|"Link,l=5"))
[6] ((dim=10|id=541|"Link,l=5"), (dim=3|id=427|"S=1,Site,n=6"), (dim=10|id=952|"Link,l=6"))
[7] ((dim=10|id=952|"Link,l=6"), (dim=3|id=909|"S=1,Site,n=7"), (dim=10|id=998|"Link,l=7"))
[8] ((dim=10|id=998|"Link,l=7"), (dim=3|id=334|"S=1,Site,n=8"), (dim=10|id=40|"Link,l=8"))
[9] ((dim=10|id=40|"Link,l=8"), (dim=3|id=947|"S=1,Site,n=9"), (dim=10|id=470|"Link,l=9"))
[10] ((dim=10|id=470|"Link,l=9"), (dim=3|id=611|"S=1,Site,n=10"), (dim=9|id=760|"Link,l=10"))
[11] ((dim=9|id=760|"Link,l=10"), (dim=3|id=564|"S=1,Site,n=11"), (dim=3|id=642|"Link,l=11"))
[12] ((dim=3|id=642|"Link,l=11"), (dim=3|id=700|"S=1,Site,n=12"))

julia> h5open("testfilev0.2.h5", "w") do file 
              write(file, "psi0v0.2", psi0)
              end

julia> file02 = h5open("testfilev0.2.h5","r");

julia> read(file02,"psi0v0.2")
Dict{String, Any} with 15 entries:
  "MPS[9]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[10]" => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[1]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>2, "index_1"=>Dict{String, Any}(…
  "MPS[3]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[4]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[11]" => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[6]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "length"  => 12
  "MPS[12]" => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>2, "index_1"=>Dict{String, Any}(…
  "MPS[8]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "llim"    => 0
  "MPS[2]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[7]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "MPS[5]"  => Dict{String, Any}("inds"=>Dict{String, Any}("length"=>3, "index_1"=>Dict{String, Any}(…
  "rlim"    => 2

julia> read(file02,"psi0v0.2")["MPS[1]"]
Dict{String, Any} with 2 entries:
  "inds"    => Dict{String, Any}("length"=>2, "index_1"=>Dict{String, Any}("dim"=>3, "id"=>0xcd39a7331b4c6bcc, "plev"=>0, "tags"=>Dict{…
  "storage" => Dict{String, Any}("data"=>[0.0969137, -0.108583, 0.137754, -0.369503, -0.230657, -0.0747525, -0.0144793, -0.36735, -0.25…
```
Above is the behaviour of v0.2 and here below is the HDF5 dictionary entry of an MPSv0.1.41:
```julia
julia> read(file01,"psi0v0.1")["MPS[1]"]
Dict{String, Any} with 2 entries:
  "inds"  => Dict{String, Any}("length"=>2, "index_1"=>Dict{String, Any}("dim"=>3, "id"=>0xd299becb1442fb36, "plev"=>0, "tags"=>Dict{St…
  "store" => Dict{String, Any}("data"=>[-0.139294, 0.135728, 0.21705, 0.15732, -0.0650376, -0.172197, -0.036096, 0.0401034, -0.104529, …
```

As you can see, there are a couple of different dictionaries defined that contain the entire information.
The MPS is saved as a dictionary. Each ITensor composing the MPS is also a dictionary. The dictionary key for the data inside an ITensor (e.g. key `MPS[1]"` of the MPS) changes name from `store` in v0.1 to `storage` in v0.2.

Could the functionthat defines how to read an ITensor be fixed to read either `store` or `storage`?

**Comment by mtfishman** [comment]
Thanks for the in-depth example! That is very helpful.

Indeed, I do see that I changed the name of the field that gets saved from `"store"` in v0.1 to `"storage"` in v0.2, which was probably a mistake (I think just from doing a general find and replace, since we changed the name of the field within the type).

https://github.com/ITensor/ITensors.jl/blob/cbbe909dcbda4cf70c8ecb7679128962cda3ebc4/src/itensor.jl#L1991
https://github.com/ITensor/ITensors.jl/blob/71c19d63234b7771530eaf5018d489da2f48b5ae/src/itensor.jl#L2570

I just confirmed that with a more minimal example of just writing and reading an ITensor, I get a similar error. I also just tested that when I change `"storage"` back to `"store"` in the ITensor `read` function it works now, I'll make a PR.
